### PR TITLE
[WIP] Authentication front and backend

### PR DESCRIPTION
**Pull request by MitchellSlavik**

### CODE REVIEW

File: zoo_api/server/boot/root.js [hunk 1]
```diff
@@ -1,7 +1,39 @@
+const bodyParser = require('body-parser');
+
+const jsonParser = bodyParser.json();
+
+const loginFunc = (Account, Role) => (req, res) => {
```
Comment (ColinKrist) [inline zoo_api/server/boot/root.js]: it'd be nice to have messages regarding missing login info before a login request is made.
```suggestion
const loginFunc = (Account, Role) => (req, res) => {
  if(!req.body.email){
    res.status(401).json({ message: 'Email address missing. Please enter email address.', status: 401 });
  } else if(!req.body.password) {
    res.status(401).json({ message: 'Password missing. Please enter password.', status: 401 });
  } else {
  // do the rest of the account login
```

File: zoo_api/server/boot/root.js [hunk 2]
```diff
@@ -1,7 +1,39 @@
...
+  }, (err, accessToken) => {
+    if (err) {
+      res.status(401).json({ message: 'Unauthorized', status: 401 });
+    } else {
+      Role.getRoles({
+        principleType: 'account',
+        principleId: accessToken.userId,
+      }, (err2, roles) => {
+        if (err2) {
+          res.status(401).json({ message: 'Unable to get roles', status: 401 });
```
Comment (ColinKrist) [inline zoo_api/server/boot/root.js]:
```suggestion
          res.status(401).json({ message: 'Unable to get role. Please contact your system administrator to assign a role to your account.', status: 401 });
```

**Review by ColinKrist — CHANGES_REQUESTED** [review]
not sure if syntax is correct on these suggestions, but I think having a specific email / password check on API side would be nice. Thinking long term here if people were to use the login for an external service having that check is good here.

**Review by ColinKrist — APPROVED** [review]

File: zoo_frontend/pages/_app.jsx [hunk 3]
```diff
@@ -4,14 +4,68 @@ import Head from 'next/head';
...
+  }
+};
 
 class MyApp extends App {
-  constructor() {
-    super();
+  static async getInitialProps({ ctx, Component }) {
+    let pageProps = {};
+    const c = cookies(ctx);
+    this.api = new Api(c.authToken || '');
```
Comment (MitchellSlavik) [inline zoo_frontend/pages/_app.jsx]: there is no 'this' in a static function

**Pull request merged**